### PR TITLE
Optimize outbox poll path: push predicates to DB, add adaptive backoff (#942)

### LINKED ISSUE #942: Outbox query performance pass
## Summary

Rewrite the hot paths in `OutboxRepository` so predicates are evaluated at the database, not in Python. Reduce per-status counting to a single fast `COUNT` per status (no projection wrapper). Refactor cleanup methods to delete directly. Add adaptive polling backoff to `OutboxProcessor` so idle-period traffic drops to near-zero.

## Problem

Four issues on the same hot path compound to make the outbox poll more expensive than it should be.

### 1. 3× over-fetch in `find_unprocessed`

[`OutboxRepository.find_unprocessed`](../blob/main/src/protean/utils/outbox.py#L316-L369) fetches `min(limit * 3, 1000)` rows from the database, then filters them in Python via `is_ready_for_processing()`:

```python
query_limit = min(limit * 3, 1000)
results = self._apply_limit_and_execute(query, query_limit)
ready_messages = [msg for msg in results if msg.is_ready_for_processing()]
ready_messages = ready_messages[:limit]
```

`is_ready_for_processing()` evaluates four conditions:

1. `status IN ('pending','failed')` — already in SQL.
2. `locked_until IS NULL OR locked_until < now` — in Python.
3. `next_retry_at IS NULL OR next_retry_at <= now` — in Python.
4. `retry_count < max_retries` — in Python (column-to-column).

The 3× factor exists to compensate for predicates 2-4 filtering rows out in Python. Pushing 2-3 into SQL eliminates the over-fetch entirely. Predicate 4 needs F()-style column-to-column comparison (separate sub-issue); in practice it almost never fires on a live workload (only when retries are exhausted) so leaving it in Python until F() lands does not reintroduce meaningful over-fetch.

### 2. `count_by_status` issues N queries with subquery wrappers

[`OutboxRepository.count_by_status`](../blob/main/src/protean/utils/outbox.py#L620-L632) loops over each `OutboxStatus` and runs `.filter(status=…).limit(1).all()` to read `result.total`. Five separate queries, each shaped like `SELECT count(*) FROM (SELECT … FROM outbox WHERE status=…) AS anon_1` — the wrapper subquery exists because the DAO computes `total` by re-wrapping the query, listing every projected column. The optimizer usually elides the projection but parser/ORM/plan-compilation cost is real and per-call.

The DAO has no first-class `count()` method that would issue `SELECT COUNT(*) FROM outbox WHERE …` directly.

### 3. Cleanup methods read-then-delete

Both [`cleanup_old_published`](../blob/main/src/protean/utils/outbox.py#L634-L656) and [`cleanup_old_abandoned`](../blob/main/src/protean/utils/outbox.py#L658-L683) fetch every matching row in full (including the `data` and `metadata_` JSON blobs), call `len()` on the result, then issue a separate `_delete_all`. Two passes over the same data, the first of which materializes blobs that will be discarded.

`_delete_all` already returns rowcount on all three adapters, so the read pass is purely wasteful.

### 4. Constant-rate polling

[`OutboxProcessor`](../blob/main/src/protean/server/outbox_processor.py#L25-L102) polls at a fixed `tick_interval` (default 1s) regardless of activity. Idle deployments still issue 86,400 polls per worker per day.

### A. Add `__isnull` lookup to Q

The Q lookup machinery (`src/protean/utils/query.py`) does not currently support `__isnull`. Add it as an additive lookup, registered for SQLAlchemy, Memory, and Elasticsearch adapters.

```python
Q(locked_until__isnull=True)   # IS NULL
Q(locked_until__isnull=False)  # IS NOT NULL
```

### B. Push 3 of 4 `is_ready_for_processing` predicates into SQL

```python
def find_unprocessed(self, limit=None, target_broker=None):
    now = datetime.now(timezone.utc)
    query = self._dao.query.filter(
        status__in=[OutboxStatus.PENDING.value, OutboxStatus.FAILED.value]
    )
    query = query.filter(Q(locked_until__isnull=True) | Q(locked_until__lt=now))
    query = query.filter(Q(next_retry_at__isnull=True) | Q(next_retry_at__lte=now))
    if target_broker is not None:
        query = query.filter(target_broker=target_broker)
    candidates = self._apply_limit_and_execute(
        query.order_by("-priority"), limit
    )
    # Final guard: retry_count < max_retries (column-to-column).
    # Pushed to SQL once F() expressions land (separate sub-issue).
    return [msg for msg in candidates if msg.retry_count < msg.max_retries]
```

The `* 3` over-fetch factor is removed. The Python post-filter is left only for the column-to-column predicate, which is a no-op for the overwhelming majority of rows.

### C. Add `QuerySet.count()`

Add a first-class `.count()` method to `QuerySet` that emits `SELECT COUNT(*) FROM table WHERE …` directly. No projection wrapper, no `.all()`, no row materialization.

```python
class QuerySet:
    def count(self) -> int:
        ...
```

Per-adapter implementation: SQLAlchemy emits `func.count()` over the criteria; Memory counts the filtered iterable; ES uses the `_count` API.

### D. Rewrite `count_by_status`

```python
def count_by_status(self) -> dict:
    return {
        status.value: self._dao.query.filter(status=status.value).count()
        for status in OutboxStatus
    }
```

Still N queries (one per status), but each is now a flat `SELECT COUNT(*)` rather than a wrapped subquery. A real `GROUP BY` would collapse to one query but requires a broader aggregate API; out of scope here.

### E. Refactor cleanup methods

```python
def cleanup_old_published(self, older_than_hours=168):
    threshold = datetime.now(timezone.utc) - timedelta(hours=older_than_hours)
    return self._dao._delete_all(
        Q(status=OutboxStatus.PUBLISHED.value, published_at__lt=threshold)
    )
```

Single statement, returns rowcount. Same shape for `cleanup_old_abandoned`.

(Note: this does single-shot delete. Bounded-batch cleanup for very large backlogs is the separate `_delete_top` sub-issue.)

### F. Adaptive polling backoff

In `OutboxProcessor`, track consecutive empty polls. On empty: double the interval, capped at `max_tick_interval` (default 5s). On non-empty: reset to base.

```python
@dataclass
class OutboxProcessor:
    tick_interval: float = 1.0
    max_tick_interval: float = 5.0

    def __post_init__(self):
        self._current_interval = self.tick_interval

    async def tick(self):
        batch = await self.get_next_batch_of_messages()
        if batch:
            self._current_interval = self.tick_interval
            await self.process_batch(batch)
        else:
            self._current_interval = min(
                self._current_interval * 2, self.max_tick_interval
            )
```

Configurable via the existing `[server.outbox]` config section. Setting `max_tick_interval == tick_interval` preserves current behavior for users who want it.

## Alternatives considered

- **Push all four predicates into SQL now**, including `retry_count < max_retries`. Rejected — needs F() expressions, a broader framework change with its own sub-issue. Three out of four predicates eliminate >99% of the over-fetch.
- **Single `GROUP BY` query for `count_by_status`.** Rejected — requires an aggregate API on QuerySet, which is significantly more design surface than `.count()`. Five flat counts is a 10×+ improvement and gets us 90% of the win.
- **Notification-based polling (LISTEN/NOTIFY).** Considered and rejected at the epic level. Postgres-only and the polling path is acceptable once predicates are in SQL.

## Backward compatibility

Additive only.

- `__isnull` is a new lookup; existing queries unaffected.
- `QuerySet.count()` is new; existing `.limit(1).all().total` keeps working.
- `find_unprocessed` API surface unchanged; semantics tighter (returns only ready messages) but `is_ready_for_processing()` already produced this set; the only difference is fewer rows leave the database.
- `cleanup_old_*` API surface unchanged.
- Adaptive backoff bounded by `max_tick_interval` config; setting it equal to `tick_interval` keeps current behavior.

## Test plan

- `tests/utils/query/` — new tests for `__isnull` across SQLAlchemy, Memory, ES.
- `tests/core/queryset/` — tests for `.count()` across adapters.
- `tests/outbox/test_outbox_repository.py`:
  - Update `test_find_unprocessed_respects_limit` to exercise the no-over-fetch path.
  - Update `test_count_by_status_returns_correct_counts` for the new query shape.
  - Add boundary tests: just-expired lock, just-arrived retry window, lock held by another worker.
- `tests/server/test_outbox_processor.py` — deterministic tests for adaptive backoff and reset on activity.
- `tests/outbox/test_outbox_perf.py` — opt-in microbenchmark capturing query count and round-trip count for the poll path.

## Acceptance criteria

- [ ] `__isnull` lookup works on all three adapters.
- [ ] `find_unprocessed` issues a single SELECT with status, lock, and retry-window predicates in WHERE; no over-fetch factor.
- [ ] `QuerySet.count()` issues `SELECT COUNT(*)` with no projection wrapper on SQLAlchemy.
- [ ] `count_by_status` produces correct counts and no longer wraps each query in a subquery.
- [ ] `cleanup_old_published` and `cleanup_old_abandoned` issue a single DELETE and return rowcount.
- [ ] `OutboxProcessor` backs off on empty polls and resets on non-empty polls; bounds configurable.
- [ ] Existing outbox tests pass; new tests cover boundary conditions and adapter parity.

### PR DESCRIPTION
**Pull request by subhashb**
## Summary

Reworks the outbox poll hot path so predicates are evaluated at the database instead of in Python, collapses per-status counting to flat `COUNT`s, deletes directly in cleanup, and lets idle deployments back off polling. The poll runs every tick, per worker, indefinitely, so wasted work here compounds.

### Query layer (additive — SQLAlchemy, Memory, Elasticsearch)
- **`__isnull` lookup** — `Q(field__isnull=True/False)` → `IS NULL` / `IS NOT NULL`. Memory lookups gain a `null_safe` flag so `_evaluate_lookup` still evaluates the predicate when the source value is `None`.
- **`QuerySet.count()`** — backed by a new `_count` DAO primitive issuing a flat `SELECT COUNT(*)` (SQLAlchemy `func.count()`, Memory `len` over the filtered set, Elasticsearch `_count` API) with no projection wrapper or entity materialization.
- **`QuerySet.all(with_total=False)`** — threaded through `_filter` on every adapter. When the caller only needs `.items`, SQLAlchemy skips the separate wrapped `COUNT` round-trip; Memory/Elasticsearch report `total` for free.

### Outbox repository
- **`find_unprocessed`** pushes status, lock-window, and retry-window predicates into the query and drops the `* 3` over-fetch factor. Fetches via `all(with_total=False)`, so each poll issues a **single SELECT**. Only the column-to-column `retry_count < max_retries` guard remains in Python (pending F() expressions); it eliminates near-zero rows on a healthy workload.
- **`count_by_status`** issues flat `SELECT COUNT(*)` per status instead of subquery-wrapped projections.
- **`cleanup_old_published` / `cleanup_old_abandoned`** issue a single `DELETE` returning rowcount, instead of fetching every matching row (including JSON blobs) first.

### Outbox processor
- **Adaptive polling backoff** via the new `outbox.max_tick_interval` config option. Empty polls double the wait up to the cap; a non-empty batch resets to the base `tick_interval`. Defaults to `None` (constant-rate polling), preserving existing behavior unless opted in. Wired through the engine to both managed and external outbox processors.

## Backward compatibility

Additive throughout — new lookup, new methods, and optional parameters whose defaults preserve current behavior. One note for adapter authors: `BaseDAO` gains a new abstract `_count`; custom DAO subclasses must implement it (consistent with the existing `_filter`/`_create`/`_update`/`_delete_all` port primitives).

## Tests

- Cross-adapter coverage for `count()` and `__isnull` (Memory, PostgreSQL, Elasticsearch).
- Boundary and query-shape tests for `find_unprocessed`, `count_by_status`, and the cleanup methods.
- Deterministic adaptive-backoff tests plus engine wiring.
- Opt-in poll-path microbenchmark (`tests/outbox/test_outbox_perf.py`) asserting statement counts/shapes on SQLAlchemy: single SELECT for `find_unprocessed`, flat unwrapped COUNTs for `count_by_status`, a lone DELETE per cleanup method.

Verified via `protean test -c FULL`: SQLite, PostgreSQL, Elasticsearch, MessageDB, Redis, Redis PubSub, and Inline all pass. (MSSQL was not exercised locally — missing unixODBC driver — but the changes are dialect-agnostic.)

Closes #942

### CODE REVIEW

File: src/protean/adapters/repository/memory.py [hunk 1]
```diff
@@ -428,12 +433,13 @@ def compound_sort_key(item):
 
             items = sorted(items, key=compound_sort_key)
 
+        # When limit is not set, we return all items
+        returned = items[offset : offset + limit] if limit else items
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/adapters/repository/memory.py]: Good catch — fixed in d38f930d. `_filter` now slices `items[offset:]` when no limit is set, so a bare `.offset(n)` skips the leading rows instead of returning the full list. Added a core test (`test_offset_without_limit_still_applies_offset`) covering it.

File: src/protean/adapters/repository/sqlalchemy.py [hunk 2]
```diff
@@ -602,10 +604,11 @@ def _filter(
...
         try:
             items = qs.all()
-            result = ResultSet(
-                offset=offset, limit=limit, total=qs_without_limit.count(), items=items
-            )
-        except Exception as exc:
+            # Skip the separate ``SELECT COUNT(*)`` round-trip when the caller
+            # does not need the full match count (e.g. the outbox poll path).
+            total = qs_without_limit.count() if with_total else len(items)
+            result = ResultSet(offset=offset, limit=limit, total=total, items=items)
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/adapters/repository/sqlalchemy.py:618]: Done in d38f930d. The `with_total=True` path now uses a flat `func.count()` over the same criteria with `order_by(None)`, avoiding the subquery wrapper and the carried ORDER BY — same shape as the `_count` primitive this PR adds.

File: tests/outbox/test_outbox_perf.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
+"""Opt-in microbenchmark for the outbox poll-path query/round-trip counts (#942).
+
+Gated behind ``@pytest.mark.database`` so it runs only against a real SQL
+provider (e.g. ``protean test --sqlite`` / ``--postgresql``), where emitted
+statements are observable and meaningful. The test attaches a SQLAlchemy
+``after_cursor_execute`` listener to count statements issued on the poll path
+and asserts the performance contract introduced by #942:
```
[the remark this answers is not in the document]
Comment (subhashb) [reply tests/outbox/test_outbox_perf.py:7]: Fixed in d38f930d — docstring now says `before_cursor_execute`, matching `capture_statements()`.
